node-lib/src/runner.rs

```diff
@@ -124,35 +127,49 @@ pub async fn initialize(
             rpc_config.cookie_file.as_deref(),
         )?;
         // TODO: get rid of the unwrap_or() after fixing the issue in #446
-        let _rpc = manager.add_subsystem(
-            "rpc",
-            rpc::Builder::new(rpc_config.into(), Some(rpc_creds))
-                .register(crate::rpc::init(
-                    manager.make_shutdown_trigger(),
-                    chain_config,
-                ))
-                .register(block_prod.into_rpc())
-                .register(chainstate.clone().into_rpc())
-                .register(mempool.into_rpc())
-                .register(p2p.clone().into_rpc())
-                .build()
-                .await?,
-        );
+        let rpc = rpc::Builder::new(rpc_config.into(), Some(rpc_creds))
+            .register(crate::rpc::init(
+                manager.make_shutdown_trigger(),
+                chain_config,
+            ))
+            .register(block_prod.clone().into_rpc())
+            .register(chainstate.clone().into_rpc())
+            .register(mempool.clone().into_rpc())
+            .register(p2p.clone().into_rpc())
+            .build();
+        let rpc = rpc.await;
+        let _rpc = manager.add_subsystem("rpc", rpc?);
+    }
+
+    if let Some(sender) = node_controller {
+        let controller = NodeController {
+            shutdown_trigger: manager.make_shutdown_trigger(),
+            chainstate: chainstate.clone(),
+            block_prod: block_prod.clone(),
+            mempool: mempool.clone(),
+            p2p: p2p.clone(),
+        };
+        sender.send(controller).expect("RemoteController channel closed");
     }
 
     Ok(manager)
 }
 
 /// Processes options and potentially runs the node.
-pub async fn run(options: Options) -> Result<()> {
-    match options.command {
+pub async fn run(
+    options: Options,
+    node_controller_sender: Option<oneshot::Sender<NodeController>>,
+) -> Result<subsystem::Manager> {
```

Comment:
It seems to me `run` is no longer an appropriate name here, since it does not invoke `Manager::main` but returns the manager instead. Maybe something like `setup`?

Reply:
That's fair. Done.